diff --git a/src/game/Level1.cpp b/src/game/Level1.cpp
index 48be118..4f81b71 100644
--- a/src/game/Level1.cpp
+++ b/src/game/Level1.cpp
@@ -1200,7 +1200,7 @@ bool ChatHandler::HandleModifyScaleCommand(const char* args)
         return false;
 
     float Scale = (float)atof((char*)args);
-    if (Scale > 3.0f || Scale <= 0.0f)
+    if (Scale > 15.0f || Scale <= 0.0f)
     {
         SendSysMessage(LANG_BAD_VALUE);
         SetSentErrorMessage(true);
