diff --git a/src/game/Unit.cpp b/src/game/Unit.cpp
index 46f5a15..6eeaa0b 100644
--- a/src/game/Unit.cpp
+++ b/src/game/Unit.cpp
@@ -2805,7 +2805,7 @@ SpellMissInfo Unit::SpellHitResult(Unit *pVictim, SpellEntry const *spell, bool
 
     // All positive spells can`t miss
     // TODO: client not show miss log for this spells - so need find info for this in dbc and use it!
-    if (IsPositiveSpell(spell->Id))
+    if (IsPositiveSpell(spell->Id)&&(!IsHostileTo(pVictim)))
         return SPELL_MISS_NONE;
 
     // Check for immune (use charges)
