diff --git a/src/shared/Database/QueryResult.h b/src/shared/Database/QueryResult.h
index 4b76b30..ca6d119 100644
--- a/src/shared/Database/QueryResult.h
+++ b/src/shared/Database/QueryResult.h
@@ -33,7 +33,7 @@ class MANGOS_DLL_SPEC QueryResult
 
         uint32 GetField_idx(const std::string &name) const
         {
-            for(FieldNames::const_iterator iter = GetFiedNames().begin(); iter != GetFiedNames().end(); ++iter)
+            for(FieldNames::const_iterator iter = GetFieldNames().begin(); iter != GetFieldNames().end(); ++iter)
             {
                 if(iter->second == name)
                     return iter->first;
@@ -53,7 +53,7 @@ class MANGOS_DLL_SPEC QueryResult
 
         uint32 GetFieldCount() const { return mFieldCount; }
         uint64 GetRowCount() const { return mRowCount; }
-        FieldNames const& GetFiedNames() const {return mFieldNames; }
+        FieldNames const& GetFieldNames() const {return mFieldNames; }
 
     protected:
         Field *mCurrentRow;
