diff --git a/src/shared/vmap/AABSPTree.h b/src/shared/vmap/AABSPTree.h
index 7a5488e..c94bfca 100644
--- a/src/shared/vmap/AABSPTree.h
+++ b/src/shared/vmap/AABSPTree.h
@@ -1549,6 +1549,10 @@ public:
     private:
         friend class AABSPTree<T>;
 
+        // Note: this is a Table iterator, we are currently defining
+        // Set iterator
+        typename Table<Member, Node*>::Iterator it;
+
         Iterator(const typename Table<Member, Node*>::Iterator& it) : it(it) {}
 
     public:
