Index: C:/ScriptDev2/scripts/zone/zulaman/boss_janalai.cpp
===================================================================
--- C:/ScriptDev2/scripts/zone/zulaman/boss_janalai.cpp	(revision 725)
+++ C:/ScriptDev2/scripts/zone/zulaman/boss_janalai.cpp	(revision 726)
@@ -16,7 +16,7 @@
 
 /* ScriptData
 SDName: Boss_Janalai
-SD%Complete: 100
+SD%Complete: 75
 SDComment:
 SDCategory: Zul'Aman
 EndScriptData */
@@ -24,6 +24,17 @@
 #include "precompiled.h"
 #include "def_zulaman.h"
 
+#define SAY_AGGRO                   -1568000
+#define SAY_FIRE_BOMBS              -1568001
+#define SAY_SUMMON_HATCHER          -1568002
+#define SAY_ALL_EGGS                -1568003
+#define SAY_BERSERK                 -1568004
+#define SAY_SLAY_1                  -1568005
+#define SAY_SLAY_2                  -1568006
+#define SAY_DEATH                   -1568007
+#define SAY_EVENT_STRANGERS         -1568008
+#define SAY_EVENT_FRIENDS           -1568009
+
 // Jan'alai
 // --Spell
 #define SPELL_FLAME_BREATH          43140
@@ -39,33 +50,6 @@
 #define SPELL_FIRE_BOMB_DUMMY       42629
 #define SPELL_FIRE_BOMB_DAMAGE      42630
 
-// -- SAYs
-#define SOUND_AGGRO                 12031
-#define SAY_AGGRO                   "Spirits of da wind be your doom!"
-#define SOUND_FIRE_BOMBS            12032
-#define SAY_FIRE_BOMBS              "I burn ya now!"
-#define SOUND_SUMMON_HATCHER        12033
-#define SAY_SUMMON_HATCHER          "Where ma hatcha? Get to work on dem eggs!"
-#define SOUND_ALL_EGGS              12034
-#define SAY_ALL_EGGS                "I show you strength... in numbers."
-#define SOUND_BERSERK               12035
-#define SAY_BERSERK                 "You done run outta time!"
-
-#define SOUND_SLAY_1                12036
-#define SAY_SLAY_1                  "It all be over now, mon!"
-#define SOUND_SLAY_2                12037
-#define SAY_SLAY_2                  "Tazaga-choo!"
-
-#define SOUND_DEATH                 12038
-#define SAY_DEATH                   "Zul'jin... got a surprise for you..."
-
-#define SOUND_AGGRO_1               12039                   //NOT USED need more information
-                                                            //NOT USED need more information (random say before aggro?)
-#define SAY_AGGRO_1                 "Come, strangers. The spirit of the dragonhawk hot be hungry for worthy souls."
-#define SOUND_AGGRO_2               12040                   //NOT USED need more information
-                                                            //NOT USED need more information (random say before aggro?)
-#define SAY_AGGRO_2                 "Come, friends. Your bodies gonna feed ma hatchlings, and your souls are going to feed me with power!"
-
 // --Summons
 #define MOB_AMANI_HATCHER           23818
 #define MOB_HATCHLING               23598
@@ -141,8 +125,8 @@
 
     void Reset()
     {
-        if(pInstance)
-            pInstance->SetData(DATA_JANALAIEVENT, NOT_STARTED);
+        if (pInstance)
+            pInstance->SetData(TYPE_JANALAIEVENT, NOT_STARTED);
 
         fire_breath_timer = 8000;
         bomb_timer = 30000;
@@ -166,78 +150,70 @@
 
     void JustDied(Unit* Killer)
     {
-        DoYell(SAY_DEATH, LANG_UNIVERSAL, NULL);
-        DoPlaySoundToSet(m_creature,SOUND_DEATH);
+        DoScriptText(SAY_DEATH, m_creature);
 
-        if(pInstance)
-            pInstance->SetData(DATA_JANALAIEVENT, DONE);
+        if (pInstance)
+            pInstance->SetData(TYPE_JANALAIEVENT, DONE);
     }
 
     void KilledUnit(Unit* victim)
     {
         switch(rand()%2)
         {
-            case 0:
-                DoYell(SAY_SLAY_1, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature,SOUND_SLAY_1);
-                break;
-            case 1:
-                DoYell(SAY_SLAY_2, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature,SOUND_SLAY_1);
-                break;
+            case 0: DoScriptText(SAY_SLAY_1, m_creature); break;
+            case 1: DoScriptText(SAY_SLAY_2, m_creature); break;
         }
     }
 
     void Aggro(Unit *who)
     {
-        if(pInstance)
-            pInstance->SetData(DATA_JANALAIEVENT, IN_PROGRESS);
+        DoScriptText(SAY_AGGRO, m_creature);
 
-        DoYell(SAY_AGGRO, LANG_UNIVERSAL, NULL);
-        DoPlaySoundToSet(m_creature,SOUND_AGGRO);
+        if (pInstance)
+            pInstance->SetData(TYPE_JANALAIEVENT, IN_PROGRESS);
     }
 
     void FireWall()                                         // Create Firewall
     {
         Creature* wall = NULL;
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[0][0],FireWallCoords[0][1],FireWallCoords[0][2],FireWallCoords[0][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[0][0],FireWallCoords[0][1]+5,FireWallCoords[0][2],FireWallCoords[0][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[0][0],FireWallCoords[0][1]-5,FireWallCoords[0][2],FireWallCoords[0][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[1][0]-2,FireWallCoords[1][1]-2,FireWallCoords[1][2],FireWallCoords[1][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[1][0]+2,FireWallCoords[1][1]+2,FireWallCoords[1][2],FireWallCoords[1][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[2][0],FireWallCoords[2][1],FireWallCoords[2][2],FireWallCoords[2][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[2][0],FireWallCoords[2][1]-5,FireWallCoords[2][2],FireWallCoords[2][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[2][0],FireWallCoords[2][1]+5,FireWallCoords[2][2],FireWallCoords[2][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[3][0]-2,FireWallCoords[3][1],FireWallCoords[3][2],FireWallCoords[3][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
 
         wall = m_creature->SummonCreature(MOB_FIRE_BOMB,FireWallCoords[3][0]+2,FireWallCoords[3][1],FireWallCoords[3][2],FireWallCoords[3][3],TEMPSUMMON_TIMED_DESPAWN,11500);
-        if(wall)
+        if (wall)
             wall->CastSpell(wall,SPELL_FIRE_WALL,false);
     }
 
@@ -251,12 +227,12 @@
             dy = (rand()%(area_dy))-(area_dy/2);
 
             Creature* bomb = DoSpawnCreature(MOB_FIRE_BOMB, dx, dy, 0, 0, TEMPSUMMON_TIMED_DESPAWN, 13000);
-            if(bomb)
+            if (bomb)
                 FireBombGUIDs[i] = bomb->GetGUID();
         }
 
         Creature* ThrowController = DoSpawnCreature(MOB_FIRE_BOMB,0,0,1,0,TEMPSUMMON_TIMED_DESPAWN,10000);
-        if(ThrowController)
+        if (ThrowController)
             ThrowControllerGUID = ThrowController->GetGUID();
 
         bombcounter = 0;
@@ -268,13 +244,13 @@
         {
             Unit* ThrowController = NULL;
             Unit* FireBomb = NULL;
-            if(ThrowControllerGUID)
+            if (ThrowControllerGUID)
                 ThrowController = Unit::GetUnit((*m_creature), ThrowControllerGUID);
 
-            if(FireBombGUIDs[bombcounter])
+            if (FireBombGUIDs[bombcounter])
                 FireBomb = Unit::GetUnit((*m_creature), FireBombGUIDs[bombcounter]);
 
-            if(ThrowController && FireBomb)
+            if (ThrowController && FireBomb)
             {
                 ThrowController->CastSpell(FireBomb,SPELL_FIRE_BOMB_THROW,true);
                 FireBomb->CastSpell(FireBomb,SPELL_FIRE_BOMB_DUMMY,false);
@@ -288,26 +264,25 @@
         if (!m_creature->SelectHostilTarget() || !m_creature->getVictim())
             return;
 
-        if(!bombing)                                        // every Spell if NOT Bombing
+        if (!bombing)                                        // every Spell if NOT Bombing
         {
             //FIRE BREATH  several videos says every 8Secounds
-            if(fire_breath_timer < diff)
+            if (fire_breath_timer < diff)
             {
-                Unit* target = NULL;
-                target = SelectUnit(SELECT_TARGET_RANDOM,0);
-                if(target)
+                if (Unit* target = SelectUnit(SELECT_TARGET_RANDOM,0))
                     DoCast(target,SPELL_FLAME_BREATH);
                 fire_breath_timer = 8000;
             }else fire_breath_timer -=diff;
 
-            if(bomb_timer < diff)
+            if (bomb_timer < diff)
             {
+                DoScriptText(SAY_FIRE_BOMBS, m_creature);
+
                 FireWall();
                 bomb_timer = 20000+rand()%20000;
                 m_creature->Relocate(JanalainPos[0][0],JanalainPos[0][1],JanalainPos[0][2],0);
                 m_creature->SendMonsterMove(JanalainPos[0][0], JanalainPos[0][1],JanalainPos[0][2],0,0,100);
-                DoYell(SAY_FIRE_BOMBS, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature,SOUND_FIRE_BOMBS);
+
                 throwBombs();
                 bombing = true;
 
@@ -320,6 +295,7 @@
                     if (Temp && m_creature->GetDistance(Temp) > 30.0)
                         DoTeleportPlayer(Temp, m_creature->GetPositionX(), m_creature->GetPositionY(), m_creature->GetPositionZ(), 0);
                 }
+
                 DoCast(m_creature,SPELL_TELETOCENTER,true); // only Effect Spell
                 DoCast(m_creature,SPELL_FIRE_BOMB_CHANNEL,false);
                 finishedbomb_timer = 11000;
@@ -333,10 +309,10 @@
             }
 
             //Enrage but only if not bombing
-            if(enragetime && !enraged)
+            if (enragetime && !enraged)
             {
-                DoYell(SAY_BERSERK, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature,SOUND_BERSERK);
+                DoScriptText(SAY_BERSERK, m_creature);
+
                 m_creature->InterruptNonMeleeSpells(false);
                 DoCast(m_creature,SPELL_ENRAGE);
                 enraged = true;
@@ -344,24 +320,24 @@
         }
 
         //Enrage after 5 minutes
-        if(enrage_timer < diff)
+        if (enrage_timer < diff)
         {
             enragetime = true;
             enrage_timer = 600000;
         }else enrage_timer -=diff;
 
-        if(bombing)                                         // every Spell if Bombing
+        if (bombing)                                        // every Spell if Bombing
         {
-            if(bombcounter < 40)
+            if (bombcounter < 40)
             {
-                if(throw_timer < diff)
+                if (throw_timer < diff)
                 {
                     throw5Bombs();
                     throw_timer = 1000;
                 }else throw_timer -=diff;
             }
 
-            if(finishedbomb_timer < diff)
+            if (finishedbomb_timer < diff)
             {
                 bombing = false;
                 finishedbomb_timer = 6000;
@@ -371,21 +347,21 @@
         }
 
         //Call Hatcher
-        if(!noeggs)
+        if (!noeggs)
         {
-            if(hatchertime < diff)
+            if (hatchertime < diff)
             {
-                if(pInstance->GetData(DATA_J_EGGSLEFT)>0 || pInstance->GetData(DATA_J_EGGSRIGHT) > 0)
+                if (pInstance->GetData(DATA_J_EGGSLEFT)>0 || pInstance->GetData(DATA_J_EGGSRIGHT) > 0)
                 {
+                    DoScriptText(SAY_SUMMON_HATCHER, m_creature);
+
                     Unit* hatcher = NULL;
-                    DoYell(SAY_SUMMON_HATCHER, LANG_UNIVERSAL, NULL);
-                    DoPlaySoundToSet(m_creature,SOUND_SUMMON_HATCHER);
                     hatcher = m_creature->SummonCreature(MOB_AMANI_HATCHER,hatcherway_l[0][0],hatcherway_l[0][1],hatcherway_l[0][2],0,TEMPSUMMON_CORPSE_TIMED_DESPAWN,10000);
-                    if(hatcher)
+                    if (hatcher)
                         hatcher->GetMotionMaster()->MovePoint(0, hatcherway_l[0][0], hatcherway_l[0][1], hatcherway_l[0][2]);
 
                     hatcher = m_creature->SummonCreature(MOB_AMANI_HATCHER,hatcherway_r[0][0],hatcherway_r[0][1],hatcherway_r[0][2],0,TEMPSUMMON_CORPSE_TIMED_DESPAWN,10000);
-                    if(hatcher)
+                    if (hatcher)
                         hatcher->GetMotionMaster()->MovePoint(0, hatcherway_r[0][0],hatcherway_r[0][1],hatcherway_r[0][2]);
                     hatchertime = 45000;
                 }
@@ -397,22 +373,19 @@
         }
 
         //WIPE after 10 minutes
-        if(wipetimer < diff)
+        if (wipetimer < diff)
         {
-            DoYell(SAY_BERSERK, LANG_UNIVERSAL, NULL);
-            DoPlaySoundToSet(m_creature,SOUND_BERSERK);
+            DoScriptText(SAY_BERSERK, m_creature);
             DoCast(m_creature,SPELL_ENRAGE);
-
             wipetimer = 30000;
         }else wipetimer -=diff;
 
         //Hatch All
-        if(!noeggs && (m_creature->GetHealth()*100) / m_creature->GetMaxHealth() < 35)
+        if (!noeggs && (m_creature->GetHealth()*100) / m_creature->GetMaxHealth() < 35)
         {
-            DoYell(SAY_ALL_EGGS , LANG_UNIVERSAL, NULL);
-            DoPlaySoundToSet(m_creature,SOUND_BERSERK);
+            DoScriptText(SAY_ALL_EGGS, m_creature);
 
-            if(pInstance)
+            if (pInstance)
                 eggs = pInstance->GetData(DATA_J_EGGSLEFT);
 
             int i;
@@ -422,11 +395,11 @@
                 int s = (rand()%20 - 10);
                 m_creature->SummonCreature(MOB_HATCHLING,JanalainPos[0][0]+s,JanalainPos[0][1]+r,JanalainPos[0][2],0,TEMPSUMMON_CORPSE_TIMED_DESPAWN,15000);
 
-                if(pInstance)
+                if (pInstance)
                     pInstance->SetData(DATA_J_HATCHLEFT,1);
             }
 
-            if(pInstance)
+            if (pInstance)
                 eggs = pInstance->GetData(DATA_J_EGGSRIGHT);
 
             for(i=1;i<=eggs;i=i+1)
@@ -434,7 +407,7 @@
                 int r = (rand()%20 - 10);
                 int s = (rand()%20 - 10);
                 m_creature->SummonCreature(MOB_HATCHLING,JanalainPos[0][0]+s,JanalainPos[0][1]+r,JanalainPos[0][2],0,TEMPSUMMON_CORPSE_TIMED_DESPAWN,15000);
-                if(pInstance)
+                if (pInstance)
                     pInstance->SetData(DATA_J_HATCHRIGHT,1);
             }
 
@@ -442,9 +415,9 @@
         }
 
         //check for reset ... exploit preventing ... pulled from his podest
-        if(reset_timer < diff)
+        if (reset_timer < diff)
         {
-            if(m_creature->GetPositionX() < -70 || m_creature->GetPositionX() > 0)
+            if (m_creature->GetPositionX() < -70 || m_creature->GetPositionX() > 0)
             {
                 EnterEvadeMode();
                 reset_timer = 5000;                         //every 5 Seca
@@ -479,7 +452,7 @@
 
     void UpdateAI(const uint32 diff)
     {
-        if(bomb_timer < diff)                               //Boom
+        if (bomb_timer < diff)                               //Boom
         {
             m_creature->CastSpell(m_creature,SPELL_FIRE_BOMB_DAMAGE,false);
             bomb_timer = 1800000;
@@ -523,11 +496,6 @@
         hatchlings = 0;
         eggs = 0;
         delete_timer = 10000;
-
-        //m_creature->RemoveAllAuras();
-        //m_creature->DeleteThreatList();
-        //m_creature->CombatStop();
-        //DoGoHome();
     }
 
     void Aggro(Unit* who)
@@ -536,20 +504,20 @@
 
     void UpdateAI(const uint32 diff)
     {
-        if(pInstance && (pInstance->GetData(DATA_JANALAIEVENT) == 1))
+        if (pInstance && (pInstance->GetData(TYPE_JANALAIEVENT) == IN_PROGRESS))
         {
-            if(!start && !hatch)
+            if (!start && !hatch)
             {
                 waytype = ( m_creature->GetPositionY() > 1150);
                 waypoint = 1;
                 start = true;
             }
 
-            if(start && !hatch)
+            if (start && !hatch)
             {
-                if(wait)
+                if (wait)
                 {
-                    if(waittimer < diff)
+                    if (waittimer < diff)
                     {
                         wait = false;
                         waittimer = 4000;
@@ -570,7 +538,7 @@
                 else
                 {
                     m_creature->GetMotionMaster()->Clear();
-                    if(waytype)
+                    if (waytype)
                         m_creature->GetMotionMaster()->MovePoint(0,hatcherway_l[waypoint][0],hatcherway_l[waypoint][1],hatcherway_l[waypoint][2]);
                     else
                         m_creature->GetMotionMaster()->MovePoint(0,hatcherway_r[waypoint][0],hatcherway_r[waypoint][1],hatcherway_r[waypoint][2]);
@@ -578,28 +546,28 @@
                 }
             }
 
-            if(start && hatch)
+            if (start && hatch)
             {
-                if(waittimer < diff)
+                if (waittimer < diff)
                 {
-                    if(!pInstance)
+                    if (!pInstance)
                         return;
 
                     waittimer = 4000;
                     Unit* hatchling;
 
-                    if(waytype)
+                    if (waytype)
                     {
                         eggs = pInstance->GetData(DATA_J_EGGSLEFT);
 
-                        if(eggs > 0)
+                        if (eggs > 0)
                             DoCast(m_creature,SPELL_HATCH_EGG);
 
                         int i;
                         for(i=1;i<=hatchlings;i=i+1)
                         {
                             eggs = pInstance->GetData(DATA_J_EGGSLEFT);
-                            if(eggs <= 0)
+                            if (eggs <= 0)
                             {
                                 if (waytype) waytype = false; else waytype = true;
                                 waittimer = 15000;
@@ -617,21 +585,21 @@
 
                         DoCast(m_creature,SPELL_HATCH_EGG);
 
-                        if(hatchlings < 5)
+                        if (hatchlings < 5)
                             hatchlings++;
                     }
                     else
                     {
                         eggs = pInstance->GetData(DATA_J_EGGSRIGHT);
 
-                        if(eggs > 0)
+                        if (eggs > 0)
                             DoCast(m_creature,SPELL_HATCH_EGG);
 
                         int i;
                         for(i=1;i<=hatchlings;i=i+1)
                         {
                             eggs = pInstance->GetData(DATA_J_EGGSRIGHT);
-                            if(eggs <= 0)
+                            if (eggs <= 0)
                             {
                                 if (waytype) waytype = false; else waytype = true;
                                 waittimer = 15000;
@@ -655,7 +623,7 @@
         }
         else
         {
-            if(delete_timer < diff)
+            if (delete_timer < diff)
             {
                 m_creature->DealDamage(m_creature, m_creature->GetHealth(), NULL, DIRECT_DAMAGE, SPELL_SCHOOL_MASK_NORMAL, NULL, false);
                 delete_timer = 10000;
@@ -696,18 +664,18 @@
 
     void UpdateAI(const uint32 diff)
     {
-        if(!start)
+        if (!start)
         {
-            if(m_creature->GetPositionY() > 1150)
+            if (m_creature->GetPositionY() > 1150)
                 m_creature->GetMotionMaster()->MovePoint(0, hatcherway_l[3][0]+rand()%4-2,hatcherway_l[3][1]+rand()%4-2,hatcherway_l[3][2]);
             else
                 m_creature->GetMotionMaster()->MovePoint(0,hatcherway_r[3][0]+rand()%4-2,hatcherway_r[3][1]+rand()%4-2,hatcherway_r[3][2]);
             start = true;
         }
 
-        if(delete_timer < diff && (pInstance && !(pInstance->GetData(DATA_JANALAIEVENT) == 1)))
+        if (delete_timer < diff && (pInstance && !(pInstance->GetData(TYPE_JANALAIEVENT) == IN_PROGRESS)))
         {
-            if(!(m_creature->getVictim()))
+            if (!(m_creature->getVictim()))
                 m_creature->DealDamage(m_creature, m_creature->GetHealth(), NULL, DIRECT_DAMAGE, SPELL_SCHOOL_MASK_NORMAL, NULL, false);
             delete_timer = 10000;
         }else delete_timer -=diff;
@@ -717,9 +685,7 @@
 
         if (buffer_timer < diff)
         {
-            Unit* target = NULL;
-            target = SelectUnit(SELECT_TARGET_RANDOM,0);
-            if(target)
+            if (Unit* target = SelectUnit(SELECT_TARGET_RANDOM,0))
                 DoCast(target,SPELL_FLAMEBUFFED);
 
             buffer_timer = 7000;
Index: C:/ScriptDev2/scripts/zone/zulaman/boss_nalorakk.cpp
===================================================================
--- C:/ScriptDev2/scripts/zone/zulaman/boss_nalorakk.cpp	(revision 725)
+++ C:/ScriptDev2/scripts/zone/zulaman/boss_nalorakk.cpp	(revision 726)
@@ -23,32 +23,23 @@
 
 #include "precompiled.h"
 
-//TODO: Trash Waves
+#define SAY_WAVE1_AGGRO         -1568010
+#define SAY_WAVE2_STAIR1        -1568011
+#define SAY_WAVE3_STAIR2        -1568012
+#define SAY_WAVE4_PLATFORM      -1568013
 
-//Unimplemented SoundIDs
-/*
-#define SOUND_NALORAKK_WAVE1    12066
-#define SOUND_NALORAKK_WAVE2    12067
-#define SOUND_NALORAKK_WAVE3    12068
-#define SOUND_NALORAKK_WAVE4    12069
+#define SAY_EVENT1_SACRIFICE    -1568014
+#define SAY_EVENT2_SACRIFICE    -1568015
 
-#define SOUND_NALORAKK_EVENT1   12078
-#define SOUND_NALORAKK_EVENT2   12079
-*/
+#define SAY_AGGRO               -1568016
+#define SAY_SURGE               -1568017
+#define SAY_TOBEAR              -1568018
+#define SAY_TOTROLL             -1568019
+#define SAY_BERSERK             -1568020
+#define SAY_SLAY1               -1568021
+#define SAY_SLAY2               -1568022
+#define SAY_DEATH               -1568023
 
-//General defines
-#define YELL_AGGRO              "You be dead soon enough!"
-#define SOUND_YELL_AGGRO        12070
-#define YELL_KILL_ONE           "Mua-ha-ha! Now whatchoo got to say?"
-#define SOUND_YELL_KILL_ONE     12075
-#define YELL_KILL_TWO           "Da Amani gonna rule again!"
-#define SOUND_YELL_KILL_TWO     12076
-#define YELL_DEATH              "I... be waitin' on da udda side...."
-#define SOUND_YELL_DEATH        12077
-                                                            //Never seen this being used, so just guessing from what I hear.
-#define YELL_BERSERK            "You had your chance, now it be too late!"
-#define SOUND_YELL_BERSERK      12074
-
 #define SPELL_BERSERK           45078                       //unsure, this increases damage, size and speed
 
 //Defines for Troll form
@@ -58,24 +49,21 @@
 #define SPELL_SURGE             42402
 #define SPELL_BEARFORM          42377
 
-#define YELL_SURGE              "I bring da pain!"
-#define SOUND_YELL_SURGE        12071
-
-#define YELL_SHIFTEDTOTROLL     "Make way for Nalorakk!"
-#define SOUND_YELL_TOTROLL      12073
-
 //Defines for Bear form
 #define SPELL_LACERATINGSLASH   42395
 #define SPELL_RENDFLESH         42397
 #define SPELL_DEAFENINGROAR     42398
 
-#define YELL_SHIFTEDTOBEAR      "You call on da beast, you gonna get more dan you bargain for!"
-#define SOUND_YELL_TOBEAR       12072
-
 struct MANGOS_DLL_DECL boss_nalorakkAI : public ScriptedAI
 {
-    boss_nalorakkAI(Creature *c) : ScriptedAI(c) {Reset();}
+    boss_nalorakkAI(Creature *c) : ScriptedAI(c)
+    {
+        pInstance = ((ScriptedInstance*)c->GetInstanceData());
+        Reset();
+    }
 
+    ScriptedInstance *pInstance;
+
     uint32 ChangeForm_Timer;
     uint32 BrutalSwipe_Timer;
     uint32 Mangle_Timer;
@@ -109,29 +97,21 @@
 
     void Aggro(Unit *who)
     {
-        DoYell(YELL_AGGRO, LANG_UNIVERSAL, NULL);
-        DoPlaySoundToSet(m_creature, SOUND_YELL_AGGRO);
+        DoScriptText(SAY_AGGRO, m_creature);
     }
 
     void KilledUnit(Unit* victim)
     {
         switch(rand()%2)
         {
-            case 0:
-                DoYell(YELL_KILL_ONE, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature, SOUND_YELL_KILL_ONE);
-                break;
-            case 1:
-                DoYell(YELL_KILL_TWO, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature, SOUND_YELL_KILL_TWO);
-                break;
+            case 0: DoScriptText(SAY_SLAY1, m_creature); break;
+            case 1: DoScriptText(SAY_SLAY2, m_creature); break;
         }
     }
 
     void JustDied(Unit* Killer)
     {
-        DoYell(YELL_DEATH,LANG_UNIVERSAL,NULL);
-        DoPlaySoundToSet(m_creature, SOUND_YELL_DEATH);
+        DoScriptText(SAY_DEATH, m_creature);
     }
 
     void UpdateAI(const uint32 diff)
@@ -143,9 +123,8 @@
         //Berserking
         if ((Berserk_Timer < diff) && (!Berserking))
         {
+            DoScriptText(SAY_BERSERK, m_creature);
             DoCast(m_creature, SPELL_BERSERK);
-            DoYell(YELL_BERSERK, LANG_UNIVERSAL, NULL);
-            DoPlaySoundToSet(m_creature, SOUND_YELL_BERSERK);
             Berserking = true;
         }else Berserk_Timer -= diff;
 
@@ -163,8 +142,8 @@
             //We just changed to troll form!
             if (!ChangedToTroll)
             {
-                DoYell(YELL_SHIFTEDTOTROLL, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature, SOUND_YELL_TOTROLL);
+                DoScriptText(SAY_TOTROLL, m_creature);
+
                 ChangedToTroll = true;
                 ChangedToBear = false;
                 //Reset spell timers
@@ -195,14 +174,12 @@
                 Unit *target = SelectUnit(SELECT_TARGET_RANDOM, 1);
 
                 //if there aren't other units, cast on the tank
-                if(!target)
+                if (!target)
                     target = m_creature->getVictim();
 
                 DoCast(target, SPELL_SURGE);
+                DoScriptText(SAY_SURGE, m_creature);
 
-                DoYell(YELL_SURGE, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature, SOUND_YELL_SURGE);
-
                 Surge_Timer = 15000 + rand()%17500;
             }else Surge_Timer -= diff;
 
@@ -220,8 +197,8 @@
             //We just changed to bear form!
             if (!ChangedToBear)
             {
-                DoYell(YELL_SHIFTEDTOBEAR, LANG_UNIVERSAL, NULL);
-                DoPlaySoundToSet(m_creature, SOUND_YELL_TOBEAR);
+                DoScriptText(SAY_TOBEAR, m_creature);
+
                 ChangedToBear = true;
                 ChangedToTroll = false;
                 //Reset spell timers
Index: C:/ScriptDev2/scripts/zone/zulaman/def_zulaman.h
===================================================================
--- C:/ScriptDev2/scripts/zone/zulaman/def_zulaman.h	(revision 725)
+++ C:/ScriptDev2/scripts/zone/zulaman/def_zulaman.h	(revision 726)
@@ -5,13 +5,26 @@
 #ifndef DEF_ZULAMAN_H
 #define DEF_ZULAMAN_H
 
-#define DATA_JANALAI        1
-#define DATA_JANALAIEVENT   2
-#define DATA_J_EGGSLEFT     3
-#define DATA_J_EGGSRIGHT    4
-#define DATA_J_HATCHLEFT    5
-#define DATA_J_HATCHRIGHT   6
+#define TYPE_EVENT_RUN      1
+#define TYPE_AKILZONEVENT   2
+#define TYPE_NALORAKKEVENT  3
+#define TYPE_JANALAIEVENT   4
+#define TYPE_HALAZZIEVENT   5
 
-#define TYPE_RAND_VENDOR_1  7
-#define TYPE_RAND_VENDOR_2  8
+#define TYPE_RAND_VENDOR_1  6
+#define TYPE_RAND_VENDOR_2  7
+
+#define DATA_AKILZON        8
+#define DATA_NALORAKK       9
+#define DATA_JANALAI        10
+#define DATA_HALAZZI        11
+
+#define DATA_J_EGGSLEFT     12
+#define DATA_J_EGGSRIGHT    13
+#define DATA_J_HATCHLEFT    14
+#define DATA_J_HATCHRIGHT   15
+
+#define DATA_GONG           16
+#define DATA_HEXLORD_GATE   17
+
 #endif
Index: C:/ScriptDev2/scripts/zone/zulaman/instance_zulaman.cpp
===================================================================
--- C:/ScriptDev2/scripts/zone/zulaman/instance_zulaman.cpp	(revision 725)
+++ C:/ScriptDev2/scripts/zone/zulaman/instance_zulaman.cpp	(revision 726)
@@ -16,7 +16,7 @@
 
 /* ScriptData
 SDName: Instance_Zulaman
-SD%Complete: 80
+SD%Complete: 60
 SDComment:
 SDCategory: Zul'Aman
 EndScriptData */
@@ -24,14 +24,21 @@
 #include "precompiled.h"
 #include "def_zulaman.h"
 
-#define ENCOUNTERS     1
+#define ENCOUNTERS     5
 #define RAND_VENDOR    2
 
 struct MANGOS_DLL_DECL instance_zulaman : public ScriptedInstance
 {
     instance_zulaman(Map *Map) : ScriptedInstance(Map) {Initialize();};
 
-    uint64 janalai;
+    uint64 akilzonGUID;
+    uint64 nalorakkGUID;
+    uint64 janalaiGUID;
+    uint64 halazziGUID;
+
+    uint64 StrangeGongGUID;
+    uint64 HexlordEntranceGUID;
+
     uint32 janalai_eggs_l;
     uint32 janalai_eggs_r;
 
@@ -40,7 +47,14 @@
 
     void Initialize()
     {
-        janalai = 0;
+        akilzonGUID = 0;
+        nalorakkGUID = 0;
+        janalaiGUID = 0;
+        halazziGUID = 0;
+
+        StrangeGongGUID = 0;
+        HexlordEntranceGUID = 0;
+
         janalai_eggs_l = 20;
         janalai_eggs_r = 20;
 
@@ -62,17 +76,39 @@
     {
         switch(creature_entry)
         {
-            case 23578:
-                janalai = creature->GetGUID();
-                break;
+            case 23574: akilzonGUID  = creature->GetGUID(); break;
+            case 23576: nalorakkGUID = creature->GetGUID(); break;
+            case 23578: janalaiGUID  = creature->GetGUID(); break;
+            case 23577: halazziGUID  = creature->GetGUID(); break;
         }
     }
 
-    uint64 GetData64(uint32 identifier)
+    void OnObjectCreate(GameObject *go)
     {
-        if(identifier  == DATA_JANALAI && janalai)
-            return janalai;
+        switch(go->GetEntry())
+        {
+            case 187359: StrangeGongGUID        = go->GetGUID(); break;
+            case 186305: HexlordEntranceGUID    = go->GetGUID(); break;
+        }
+    }
 
+    uint64 GetData64(uint32 data)
+    {
+        switch(data)
+        {
+            case DATA_AKILZON:
+                return akilzonGUID;
+            case DATA_NALORAKK:
+                return nalorakkGUID;
+            case DATA_JANALAI:
+                return janalaiGUID;
+            case DATA_HALAZZI:
+                return halazziGUID;
+            case DATA_GONG:
+                return StrangeGongGUID;
+            case DATA_HEXLORD_GATE:
+                return HexlordEntranceGUID;
+        }
         return 0;
     }
 
@@ -80,13 +116,13 @@
     {
         switch(type)
         {
-            case DATA_JANALAIEVENT:
-                if(data == 0)
+            case TYPE_JANALAIEVENT:
+                if(data == NOT_STARTED)
                 {
                     janalai_eggs_l = 20;
                     janalai_eggs_r = 20;
                 }
-                Encounters[0] = data;
+                Encounters[4] = data;
                 break;
             case DATA_J_HATCHLEFT:
                 janalai_eggs_l -= data;
@@ -100,6 +136,9 @@
             case TYPE_RAND_VENDOR_2:
                 RandVendor[1] = data;
                 break;
+            default:
+                error_log("SD2: Zul'Aman: ERROR SetData = %u for type %u does not exist/not implemented.",data,type);
+                break;
         }
     }
 
@@ -107,12 +146,22 @@
     {
         switch(type)
         {
-            case DATA_JANALAIEVENT:
+            case TYPE_EVENT_RUN:
                 return Encounters[0];
+            case TYPE_AKILZONEVENT:
+                return Encounters[1];
+            case TYPE_NALORAKKEVENT:
+                return Encounters[2];
+            case TYPE_JANALAIEVENT:
+                return Encounters[3];
+            case TYPE_HALAZZIEVENT:
+                return Encounters[4];
+
             case DATA_J_EGGSLEFT:
                 return janalai_eggs_l;
             case DATA_J_EGGSRIGHT:
                 return janalai_eggs_r;
+
             case TYPE_RAND_VENDOR_1:
                 return RandVendor[0];
             case TYPE_RAND_VENDOR_2:
Index: C:/ScriptDev2/sql/Updates/Makefile.am
===================================================================
--- C:/ScriptDev2/sql/Updates/Makefile.am	(revision 725)
+++ C:/ScriptDev2/sql/Updates/Makefile.am	(revision 726)
@@ -228,4 +228,5 @@
 	r710_scriptdev2_script_texts.sql \
 	r713_scriptdev2_script_texts.sql \
 	r715_scriptdev2_script_texts.sql \
-	r725_scriptdev2_script_texts.sql
+	r725_scriptdev2_script_texts.sql \
+	r726_scriptdev2_script_texts.sql
Index: C:/ScriptDev2/sql/Updates/r726_scriptdev2_script_texts.sql
===================================================================
--- C:/ScriptDev2/sql/Updates/r726_scriptdev2_script_texts.sql	(revision 0)
+++ C:/ScriptDev2/sql/Updates/r726_scriptdev2_script_texts.sql	(revision 726)
@@ -0,0 +1,26 @@
+DELETE FROM `script_texts` WHERE `entry` BETWEEN -1568023 AND -1568000;
+INSERT INTO `script_texts` (`entry`,`content_default`,`sound`,`type`,`language`,`comment`) VALUES
+(-1568000,'Spirits of da wind be your doom!',12031,1,0,'janalai SAY_AGGRO'),
+(-1568001,'I burn ya now!',12032,1,0,'janalai SAY_FIRE_BOMBS'),
+(-1568002,'Where ma hatcha? Get to work on dem eggs!',12033,1,0,'janalai SAY_SUMMON_HATCHER'),
+(-1568003,'I show you strength... in numbers.',12034,1,0,'janalai SAY_ALL_EGGS'),
+(-1568004,'You done run outta time!',12035,1,0,'janalai SAY_BERSERK'),
+(-1568005,'It all be over now, mon!',12036,1,0,'janalai SAY_SLAY_1'),
+(-1568006,'Tazaga-choo!',12037,1,0,'janalai SAY_SLAY_2'),
+(-1568007,'Zul\'jin... got a surprise for you...',12038,1,0,'janalai SAY_DEATH'),
+(-1568008,'Come, strangers. The spirit of the dragonhawk hot be hungry for worthy souls.',12039,1,0,'janalai SAY_EVENT_STRANGERS'),
+(-1568009,'Come, friends. Your bodies gonna feed ma hatchlings, and your souls are going to feed me with power!',12040,1,0,'janalai SAY_EVENT_FRIENDS'),
+(-1568010,'Get da move on, guards! It be killin\' time!',12066,1,0,'nalorakk SAY_WAVE1_AGGRO'),
+(-1568011,'Guards, go already! Who you more afraid of, dem... or me?',12067,1,0,'nalorakk SAY_WAVE2_STAIR1'),
+(-1568012,'Ride now! Ride out dere and bring me back some heads!',12068,1,0,'nalorakk SAY_WAVE3_STAIR2'),
+(-1568013,'I be losin\' me patience! Go on: make dem wish dey was never born!',12069,1,0,'nalorakk SAY_WAVE4_PLATFORM'),
+(-1568014,'What could be better than servin\' da bear spirit for eternity? Come closer now. Bring your souls to me!',12078,1,0,'nalorakk SAY_EVENT1_SACRIFICE'),
+(-1568015,'Don\'t be delayin\' your fate. Come to me now. I make your sacrifice quick.',12079,1,0,'nalorakk SAY_EVENT2_SACRIFICE'),
+(-1568016,'You be dead soon enough!',12070,1,0,'nalorakk SAY_AGGRO'),
+(-1568017,'I bring da pain!',12071,1,0,'nalorakk SAY_SURGE'),
+(-1568018,'You call on da beast, you gonna get more dan you bargain for!',12072,1,0,'nalorakk SAY_TOBEAR'),
+(-1568019,'Make way for Nalorakk!',12073,1,0,'nalorakk SAY_TOTROLL'),
+(-1568020,'You had your chance, now it be too late!',12074,1,0,'nalorakk SAY_BERSERK'),
+(-1568021,'Mua-ha-ha! Now whatchoo got to say?',12075,1,0,'nalorakk SAY_SLAY1'),
+(-1568022,'Da Amani gonna rule again!',12076,1,0,'nalorakk SAY_SLAY2'),
+(-1568023,'I... be waitin\' on da udda side....',12077,1,0,'nalorakk SAY_DEATH');

Property changes on: C:\ScriptDev2\sql\Updates\r726_scriptdev2_script_texts.sql
___________________________________________________________________
Name: svn:eol-style
   + native

Index: C:/ScriptDev2/sql/scriptdev2_script_texts.sql
===================================================================
--- C:/ScriptDev2/sql/scriptdev2_script_texts.sql	(revision 725)
+++ C:/ScriptDev2/sql/scriptdev2_script_texts.sql	(revision 726)
@@ -745,7 +745,33 @@
 (-1565019,'grows in size!',0,2,0,'gruul EMOTE_GROW');
 
 -- -1 568 000 ZUL'AMAN
+INSERT INTO `script_texts` (`entry`,`content_default`,`sound`,`type`,`language`,`comment`) VALUES
+(-1568000,'Spirits of da wind be your doom!',12031,1,0,'janalai SAY_AGGRO'),
+(-1568001,'I burn ya now!',12032,1,0,'janalai SAY_FIRE_BOMBS'),
+(-1568002,'Where ma hatcha? Get to work on dem eggs!',12033,1,0,'janalai SAY_SUMMON_HATCHER'),
+(-1568003,'I show you strength... in numbers.',12034,1,0,'janalai SAY_ALL_EGGS'),
+(-1568004,'You done run outta time!',12035,1,0,'janalai SAY_BERSERK'),
+(-1568005,'It all be over now, mon!',12036,1,0,'janalai SAY_SLAY_1'),
+(-1568006,'Tazaga-choo!',12037,1,0,'janalai SAY_SLAY_2'),
+(-1568007,'Zul\'jin... got a surprise for you...',12038,1,0,'janalai SAY_DEATH'),
+(-1568008,'Come, strangers. The spirit of the dragonhawk hot be hungry for worthy souls.',12039,1,0,'janalai SAY_EVENT_STRANGERS'),
+(-1568009,'Come, friends. Your bodies gonna feed ma hatchlings, and your souls are going to feed me with power!',12040,1,0,'janalai SAY_EVENT_FRIENDS'),
 
+(-1568010,'Get da move on, guards! It be killin\' time!',12066,1,0,'nalorakk SAY_WAVE1_AGGRO'),
+(-1568011,'Guards, go already! Who you more afraid of, dem... or me?',12067,1,0,'nalorakk SAY_WAVE2_STAIR1'),
+(-1568012,'Ride now! Ride out dere and bring me back some heads!',12068,1,0,'nalorakk SAY_WAVE3_STAIR2'),
+(-1568013,'I be losin\' me patience! Go on: make dem wish dey was never born!',12069,1,0,'nalorakk SAY_WAVE4_PLATFORM'),
+(-1568014,'What could be better than servin\' da bear spirit for eternity? Come closer now. Bring your souls to me!',12078,1,0,'nalorakk SAY_EVENT1_SACRIFICE'),
+(-1568015,'Don\'t be delayin\' your fate. Come to me now. I make your sacrifice quick.',12079,1,0,'nalorakk SAY_EVENT2_SACRIFICE'),
+(-1568016,'You be dead soon enough!',12070,1,0,'nalorakk SAY_AGGRO'),
+(-1568017,'I bring da pain!',12071,1,0,'nalorakk SAY_SURGE'),
+(-1568018,'You call on da beast, you gonna get more dan you bargain for!',12072,1,0,'nalorakk SAY_TOBEAR'),
+(-1568019,'Make way for Nalorakk!',12073,1,0,'nalorakk SAY_TOTROLL'),
+(-1568020,'You had your chance, now it be too late!',12074,1,0,'nalorakk SAY_BERSERK'),
+(-1568021,'Mua-ha-ha! Now whatchoo got to say?',12075,1,0,'nalorakk SAY_SLAY1'),
+(-1568022,'Da Amani gonna rule again!',12076,1,0,'nalorakk SAY_SLAY2'),
+(-1568023,'I... be waitin\' on da udda side....',12077,1,0,'nalorakk SAY_DEATH');
+
 -- -1 580 000 SUNWELL PLATEAU
 INSERT INTO `script_texts` (`entry`,`content_default`,`sound`,`type`,`language`,`comment`) VALUES
 (-1580000,'Aggh! No longer will I be a slave to Malygos! Challenge me and you will be destroyed!',12422,1,0,'kalecgos SAY_EVIL_AGGRO'),


C++ scripts converted to use database texts, some additions for future dev
* Zul'Aman