Index: scripts/zone/coilfang_resevoir/serpent_shrine/boss_lady_vashj.cpp
===================================================================
--- scripts/zone/coilfang_resevoir/serpent_shrine/boss_lady_vashj.cpp	(revision 738)
+++ scripts/zone/coilfang_resevoir/serpent_shrine/boss_lady_vashj.cpp	(working copy)
@@ -807,10 +807,19 @@
 
             //remove this item
             player->DestroyItemCount(31088, 1, true);
+            return true;
         }
-    }
+        else if( targets.getUnitTarget()->GetTypeId()==TYPEID_UNIT )
+            return false;
+        else if(targets.getUnitTarget()->GetTypeId()==TYPEID_PLAYER)
+        {
+            player->DestroyItemCount(31088, 1, true);
+            player->CastSpell(targets.getUnitTarget(), 38134, true);
+            return true;
+        }
+      }
     return true;
-}
+};
 
 CreatureAI* GetAI_boss_lady_vashj(Creature *_Creature)
 {
