Index: scripts/zone/black_temple/boss_supremus.cpp
===================================================================
--- scripts/zone/black_temple/boss_supremus.cpp	(revision 667)
+++ scripts/zone/black_temple/boss_supremus.cpp	(working copy)
@@ -115,7 +115,7 @@
         CheckTimer = 1000;
         SupremusGUID = 0;
         FireballTimer = 500;
-        GeyserTimer = 0;
+        GeyserTimer = 2000;
     }
 
     void Aggro(Unit *who) {}
@@ -332,15 +332,7 @@
 
                 if(target)
                 {
-                    Creature* Volcano = NULL;
-                    Volcano = SummonCreature(CREATURE_VOLCANO, target);
-
-                    if(Volcano)
-                    {
-                        DoCast(target, SPELL_VOLCANIC_ERUPTION);
-                        ((npc_volcanoAI*)Volcano->AI())->SetSupremusGUID(m_creature->GetGUID());
-                    }
-
+                    DoCast(target, SPELL_VOLCANIC_ERUPTION);
                     DoTextEmote("roars and the ground begins to crack open!", NULL);
                     SummonVolcanoTimer = 10000;
                 }
