Index: scripts/zone/tempest_keep/the_eye/boss_void_reaver.cpp
===================================================================
--- scripts/zone/tempest_keep/the_eye/boss_void_reaver.cpp	(revision 738)
+++ scripts/zone/tempest_keep/the_eye/boss_void_reaver.cpp	(working copy)
@@ -121,7 +121,7 @@
                 target = Unit::GetUnit(*m_creature, (*itr)->getUnitGuid());
 
                 //15 yard radius minimum
-                if (target && target->GetDistance2d(m_creature) > 15)
+                if (target && target->GetDistance2d(m_creature) > 18)
                     target_list.push_back(target);
 
                 target = NULL;
@@ -131,12 +131,7 @@
                 target = *(target_list.begin()+rand()%target_list.size());
 
             if (target)
-            {
-                Unit* Spawn = NULL;
-                Spawn = m_creature->SummonCreature(CREATURE_ORB_TARGET, target->GetPositionX(), target->GetPositionY(), target->GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN, 10000);
-                if (Spawn)
-                    m_creature->CastSpell(Spawn, SPELL_ARCANE_ORB_TRIGGER, true);
-            }
+                m_creature->CastSpell(target, SPELL_ARCANE_ORB_TRIGGER, true);
 
             ArcaneOrb_Timer = 3000;
         }else ArcaneOrb_Timer -= diff;
